### ISSUE TITLE
Passing 'server' instance to handlers

### ISSUE THREAD

**Issue by NShahri** [restify/node-restify#1595]
- [x] Used appropriate template for the issue type
- [x] Searched both open and closed issues for duplicates of this issue
- [x] Title adequately and _concisely_ reflects the feature or the bug

# Feature Request

## Use Case

At the moment developers can not use arrow functions as handler when they need 'server' instance, because when arrow function is defined, current context ('this') would be null. 

Passing server as 4th param to handler can make developer life easier. so developers can make decision which way to access to 'server' instance.

## Example API

Using arrow function causes 'this' to be null.
```
file1.js:

const handler = (req, res, next) => {console.log(this); next();}
```
```
file2.js:

server.get('/movie', handler);
```

Solution:
Pass 'server' param to handlers:
```
file1.js:

const handler = (req, res, next, server) => {console.log(server); next();}
```
```
file2.js:

server.get('/movie', handler);
```

## Are you willing and able to implement this?

"Yes"
https://github.com/restify/node-restify/pull/1596

**Comment by retrohacker** [comment]
Hey @NShahri!

Thank you for taking the time to open this issue and for including an accompanying PR.

I need some more context before I can review this. I'm not sure I fully understand the problem this is solving. Is there a reason to choose an arrow function expression over a standard function expression here? Standard function expressions give you a ` this` object which seems suited to the problem we are trying to solve.

**Comment by NShahri** [comment]
@retrohacker Yes you are right, but

- first of all it means we need to tell developer if your code has a dependency to `server` object, you can not use arrow function as your route handler.
- about automation test, in current approach we need to mock the 'server' and bind the handler to this context. 
```
movieGetHandler.bind(mockServer)(mockRequest, mockResponse)
OR
movieGetHandler(mockRequest, mockResponse, mockServer)
```
- ES6: Using static function in class, `this` is mapped to a constructor of the class which owns the static function. 

**My idea:** I can name `server`, `request`, `response` and `next` all together `context`. and we need to pass context as param to **state less** function which can accept the `context` as param.

[As sample](https://github.com/NShahri/hypermedia-boilerplate/blob/master/packages/api/routes/movies.js#L51) This is a project which I am working on.

**Comment by retrohacker** [comment]
> first of all it means we need to tell developers if your code has a dependency to server object, you can not use arrow function as your route handler.

I believe we already implicitly state this since `server` is passed as `this`.

> about automation test, in current approach we need to mock the 'server' and bind the handler to this context.

`movieGetHandler.call(mockServer, req, res, next)`

> ES6 static functions: in static method this is mapped to a constructor of the class which owns the static function. and it is not possible to use as route handler.

This is very true, and I can see how it would be a pain to work around.

> My idea: I can name server, request, response and next all together context. and we need to pass context as param to state less function which can accept the context as param.

Ah this is clever. I wonder if we could expose this as a separate API? I'm reluctant to pass parameters after the callback, as that breaks assumptions about callback functions in JS.

What are your thoughts @DonutEspresso?

**Comment by hekike** [comment]
@retrohacker just to mention: we will have a completely new middleware system in 7.x, if we decide to go with this we need to re-implement/document etc.

**Comment by NShahri** [comment]
@hekike @retrohacker Could you please tell me where I can find more info about version 7.x. How can I make proposal for new version (ex: make separation between server and router objects)

**Comment by retrohacker** [comment]
@NShahri they are all in-flight PRs on this repo :smile:

**Comment by DonutEspresso** [comment]
Hi @NShahri, thanks for opening the issue and sharing your use case! 

This happens to be one of those scenarios where the arrow functions are doing what they were designed to do. Imagine the following example, pulled from MDN:

```js
var obj = {
  i: 10,
  b: () => console.log(this.i, this),
  c: function() {
    console.log(this.i, this);
  }
}
obj.b(); // prints undefined, Window {...} (or the global object)
obj.c(); // prints 10, Object {...}
```

Arrow functions were designed intentionally to not have a `this` value, so the behavior you're observing is by design. As I understand arrow functions weren't designed to supersede all `function` declarations, but to simplify scenarios where you didn't want to lose `this`. Given that context, it looks like arrow functions have introduced a problem that didn't exist previously. What value are arrow functions providing in this scenario that standard functions are not providing?

**Comment by NShahri** [comment]
Hi @DonutEspresso , Thanks for your comment.

I agree about what you mentioned above, But in my case it is a big project which I prefer to define handlers in separated file and register them in another file. Also I am using ES6 and `babel` and `preset-env` to transpile my code based on nodejs version which I am using.

I think it is common in ES6 to implement the same as following:

handler1.js:
```javascript
export const fileHandler = (req, res, next) => {
    // .....
    next();
}
```

handler2.js
```javascript
export class MovieHandler {
    static getMoviesHandler(req, res, next) {
         // ....
         next();
    }
    static postMoviesHandler(req, res, next) {
         // ....
         next();
    }
}

server.js:
```javascript
import fileHandler from './handler1.js';
import MovieHandler from './handler2.js';

server.get({name: 'file', path: '/file'}, fileHandler);
server.get({name: 'movie', path: '/movie'}, MovieHandler.getMoviesHandler);
server.post({name: 'create-movie', path: '/movie'}, MovieHandler.postMoviesHandler);
```

I think exporting functions or a class with static methods is common in ES6, and I think it make sense to register handlers in separated files.

I completely agree about what you mentioned above, but on the other hand we can support all possible ways which developer can define handlers.

I am not sure about features in next major release, but we can use the following pattern (This is only idea and definitely it is breaking change).
 
```javascript
export function fileHandler({req, res, server, router}, next){
}
```

**Comment by retrohacker** [comment]
Hey @NShahri,

Could I suggest a wrapper for the short-term? Something along the lines of:

```js
// Defined somewhere else in source and imported/required
function wrap(fn) {
  return function wrapper(req, res, next) {
    fn(req, res, this, next);
  }
}

import fileHandler from './handler1.js';
import MovieHandler from './handler2.js';

server.get({name: 'file', path: '/file'}, wrap(fileHandler));
server.get({name: 'movie', path: '/movie'}, wrap(MovieHandler.getMoviesHandler));
server.post({name: 'create-movie', path: '/movie'}, wrap(MovieHandler.postMoviesHandler));
```

**Comment by retrohacker** [comment]
Also, thank you for continuing to engage on this issue :heart:

restify was built in such a way that we take the user provided functions and make them "part" of the server class by binding the this context. It makes sense when they are later invoked/accessed off of the server object itself. Moving forward, we may rethink this but from the conversation so far it isn't obvious what that would look like. The wrapper should allow you to move forward with your current ES6 syntax without being blocked by this discussion.
